.config/mise/tasks/test/unit.sh

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env bash
+#MISE description="Run unit tests"
+set -euo pipefail
+
+#USAGE flag "--coverage" negate="--no-coverage" default="false" help="Run unit tests with coverage, also reads `$TEST_COVERAGE`"
+#USAGE flag "--powerset" negate="--no-powerset" default="false" help="Run unit tests with powerset, also reads `$TEST_POWERSET`"
+#USAGE arg "<package>" help="The package to run unit tests for"
+#USAGE arg "[arguments]..." double_dash="required" default="" help="Additional arguments to pass to the test runner"
+
+PACKAGE=$usage_package
+COVERAGE=$usage_coverage
+POWERSET=$usage_powerset
+ARGUMENTS=$usage_arguments
+
+# Check if the package argument starts with `@rust/` if that isn't the case exit out
+if [[ $PACKAGE != "@rust/"* ]]; then
+    echo "Error: Only rust crates are supported for now"
+    exit 1
+fi
+
+# Remove the package namespace from the package to get the crate name
+CRATE=${PACKAGE#*@rust/}
+
+# Run coverage instead of unit tests if `TEST_COVERAGE` is set to `true` or `1`
+if [[ $COVERAGE == "true" || ${TEST_COVERAGE:-false} == 'true' || ${TEST_COVERAGE:-false} == '1' ]]; then
+    EXCLUSIONS=$(
+        cargo metadata --format-version=1 --no-deps \
+            | jq -r --arg crate "$CRATE" '.packages[] | select(.name != $crate) | .manifest_path | rtrimstr("/Cargo.toml") | gsub("/"; "\\/"; "g")' \
+            | paste -sd "|" -
+    )
+
+    # under CI we use LCOV
+    if [[ ${CI:-0} == "1" || ${CI:-0} == "true" ]]; then
+        REPORT_FLAGS="--lcov --output-path lcov.info"
+    else
+        REPORT_FLAGS="--html --open"
+    fi
+
+    if [[ $POWERSET == "true" || ${TEST_POWERSET:-false} == 'true' || ${TEST_POWERSET:-false} == '1' ]]; then
+        CARGO_COMMAND="hack --optional-deps --feature-powerset"
+        NEXTEST_ARGS=""
+    else
+        CARGO_COMMAND=""
+        NEXTEST_ARGS="--all-features"
+    fi
+
+    cargo llvm-cov clean --workspace
+    cargo $CARGO_COMMAND llvm-cov --ignore-filename-regex "$EXCLUSIONS" -p "$CRATE" --branch --no-report nextest $NEXTEST_ARGS --all-targets --cargo-profile coverage $ARGUMENTS
+    cargo llvm-cov --ignore-filename-regex "$EXCLUSIONS" -p "$CRATE" --branch --no-clean $REPORT_FLAGS --doctests test --all-features --profile coverage --doc
+
+    exit 0
+fi
+
+# Run unit tests with powerset if `TEST_POWERSET` is set to `true` or `1`
+if [[ $POWERSET == "true" || ${TEST_POWERSET:-false} == 'true' || ${TEST_POWERSET:-false} == '1' ]]; then
+    cargo hack --optional-deps --feature-powerset nextest run -p "$CRATE" $ARGUMENTS
+    cargo test --all-features --doc -p "$CRATE"
+    exit 0
+fi
+
+LOGFILE=$(mktemp)
+trap 'rm -f "$LOGFILE"' EXIT
+
+cargo test --all-features --doc -p "$CRATE" >"$LOGFILE" 2>&1 &
```

Comment:
I appreciate why you do that, I guess because doc-tests are used to be slow as they compile per test. But this is likely conflicting with the other test invocation, so the majority of time they'd block each other and since recently, doc tests are only slow if they fail. If they pass, they are compiled in a single binary, which greatly increases execution time.

Let's stick to the simple approach and just call doc-tests after normal tests

Reply:
compile-fail doctests (which hashql-core has) are still quite slow because of it (which is what initially motivated this approach) but you're right in that it doesn't make sense to overcomplicate the script because of it. done in [a670d59](https://github.com/hashintel/hash/pull/7167/commits/a670d59ce22415b57105182c601a51929b109c07)

Reply:
I actually did not t hink of compile-fail doc-tests. Maybe, it's possible to split these later into two profiles or so, but nothing we need to worry about now I guess. When running low-level tests such as `hashql-core` we need to test downstream packages anyway, which takes way longer to run (such as integration tests). Thank you!